crates/radix-engine-toolkit/src/models/canonical_address_types.rs

```diff
@@ -0,0 +1,342 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::utils::{
+    network_definition_from_network_id, network_id_from_address_string,
+};
+use paste::paste;
+use radix_engine_common::prelude::*;
+use scrypto::prelude::AddressBech32Decoder;
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+
+#[derive(Debug)]
+pub enum CanonicalAddressError {
+    FailedToDecodeBech32,
+    FailedToEncodeBech32,
+}
+
+// Required for serde Serializer
+impl Display for CanonicalAddressError {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "{:?}", self)
+    }
+}
+
+pub type NetworkId = u8;
+
+pub trait CanonicalAddress: FromStr + std::fmt::Display {
+    fn entity_type(&self) -> EntityType;
+    fn network_id(&self) -> NetworkId;
+    fn to_bech32(&self) -> Result<String, CanonicalAddressError>;
+}
+pub trait CanonicalAddressResourceType {
```

Comment:
I don't think that there is a need for this trait, I will go ahead and remove it.